src/main/java/org/folio/inventory/support/KafkaConsumerVerticle.java

```diff
@@ -0,0 +1,144 @@
+  private JsonObject config;
+  private HttpClient httpClient;
+  private Storage storage;
+  private MappingMetadataCache mappingMetadataCache;
+
+  @Override
+  public void stop(Promise<Void> stopPromise) {
+    List<Future<Void>> stopFutures = consumerWrappers.stream()
+      .map(KafkaConsumerWrapper::stop)
+      .toList();
+
+    GenericCompositeFuture.join(stopFutures)
+      .onComplete(ar -> stopPromise.complete());
+  }
+
+  protected abstract Logger getLogger();
+
+  protected KafkaConsumerWrapper<String, String> createConsumer(String eventType) {
+    var kafkaConsumerWrapper = KafkaConsumerWrapper.<String, String>builder()
+      .context(context)
+      .vertx(vertx)
+      .kafkaConfig(getKafkaConfig())
+      .loadLimit(getLoadLimit())
+      .globalLoadSensor(new GlobalLoadSensor())
+      .subscriptionDefinition(getSubscriptionDefinition(getKafkaConfig().getEnvId(), eventType))
+      .build();
+    consumerWrappers.add(kafkaConsumerWrapper);
+    return kafkaConsumerWrapper;
+  }
+
+  protected JsonObject getConfig() {
+    return ofNullable(config).orElseGet(() -> {
+      config = vertx.getOrCreateContext().config();
+      return config;
+    });
+  }
+
+  protected KafkaConfig getKafkaConfig() {
+    return ofNullable(kafkaConfig).orElseGet(() -> {
+      kafkaConfig = KafkaConfig.builder()
+        .envId(getConfig().getString(KAFKA_ENV))
+        .kafkaHost(getConfig().getString(KAFKA_HOST))
+        .kafkaPort(getConfig().getString(KAFKA_PORT))
+        .okapiUrl(getConfig().getString(OKAPI_URL))
+        .replicationFactor(parseInt(getConfig().getString(KAFKA_REPLICATION_FACTOR)))
+        .maxRequestSize(parseInt(getConfig().getString(KAFKA_MAX_REQUEST_SIZE)))
+        .build();
+      getLogger().info("kafkaConfig: {}", kafkaConfig);
+      return kafkaConfig;
+    });
+  }
+
+  protected HttpClient getHttpClient() {
+    return ofNullable(httpClient).orElseGet(() -> {
+      httpClient = vertx.createHttpClient();
+      return httpClient;
+    });
+  }
+
+  protected Storage getStorage() {
+    return ofNullable(storage).orElseGet(() -> {
+      storage = Storage.basedUpon(getConfig(), vertx.createHttpClient());
+      return storage;
+    });
+  }
+
+  protected MappingMetadataCache getMappingMetadataCache() {
+    return ofNullable(mappingMetadataCache).orElseGet(() -> {
+      var mappingMetadataExpirationTime = getCacheEnvVariable(getConfig(), METADATA_EXPIRATION_TIME);
+      mappingMetadataCache = new MappingMetadataCache(vertx, getHttpClient(),
+        Long.parseLong(mappingMetadataExpirationTime));
+      return mappingMetadataCache;
+    });
+  }
+
+  protected String getCacheEnvVariable(JsonObject config, String variableName) {
+    var cacheExpirationTime = config.getString(variableName);
+    if (isBlank(cacheExpirationTime)) {
+      cacheExpirationTime = CACHE_EXPIRATION_DEFAULT;
+    }
+    return cacheExpirationTime;
+  }
+
+  protected int getMaxDistributionNumber() {
```

Comment:
why so many protected methods? suggest to move to the private scope most of them

Reply:
because of inheritance. How could we make it private in case we can't extract corresponding logic to parent class?

Reply:
discussed, agreed it's fine